extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TransactionalInterceptorBase.java

```diff
@@ -86,8 +104,21 @@ protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm) thro
     protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm, RunnableWithException afterEndTransaction)
             throws Exception {
 
-        tm.begin();
-        Transaction tx = tm.getTransaction();
+        TransactionConfiguration configAnnotation = getTransactionTimeout(ic);
+        if (configAnnotation != null) {
+            tm.setTransactionTimeout(configAnnotation.timeout());
```

Comment:
So I wonder if `invokeInCallerTx` should raise a warning if `@TransactionConfiguration.timeout` is present in an inner method called within an already executed transaction.

WARNING would help catch bugs, exception would be better but I can imagine a method being reused and wanting to be used both within a tx called and as top level.
We could force people to create a new top level method to clarify the semantic

```
public class FooService {
    public nestable() { ... }

    @Transactional
    @TransactionConfig(timeout=900)
    public topLevelWithTimeout() { ... }
}

public class BarService {
  @Inject Foo foo;

  public topLevel() { return foo.nestable(); }
}
```

My gut feeling would be to start with the exception and wait for people's reaction. Best case wer help prevent bugs, worse case, we become more lenient upon outcry.

Thoughts @gsmet @michaelmusgrove @tomjenkinson

Reply:
Personally my reaction would be to just ignore TransactionConfiguration.timeout if it is not a newly started transaction.

Some other option would be trace logging for the condition in case someone wants to find out why their transaction does not have the timeout they are expecting?

Reply:
The main question here is what will occurs if we nest `@Transactional` annotated methods. As JTA didn't allow nested transactions, what will do the current implementation ? 
Reading the code I'm not sure if it's a no-op or an error.
The transaction guide said nothing about it neither : 
> The easiest way to define your transaction boundaries is to use the @Transactional annotation on your entry method 

So I will test it and see what happens ;) I can then update the guide with the current behaviour (and we can implement a warning if we want but as it is said, if the annotation is set on a library method we cannot do anything about it so maybe the best is to decide that it's a no-op).

Reply:
Is it intended to allow nested transactions? I was expecting that this worked much like CDI/JTA and the beans control a single transaction (and multiple beans participate in the same tx)?

Reply:
Yes it's not nested transaction, it's nested method calls from methods annotated with `@Transactional`. For that EJB and now CDI+Transaction have a clear semantic defined by `Transactional.TxType`.

Reply:
We have not resolved that one. Do we want to 

1. silently ignore or debug log
2. throw an exception

We need a choice now because while we can go from 2 to 1, we can't go from 1 to 2.
Thoughts @gsmet @Sanne ?

Reply:
throw an exception: the annotation shouldn't be there if its usage is unsupportable - and like you said, we can relax such a rule later.

But also: with the timeout attribute of `@TransactionConfig` being `int` it's not nullable.
This makes the annotation not extensible as it won't be possible to use it as "host" for additional attributes in future, if the user doesn't also mean to change the timetout.

If we accept that it's only meant to host the timeout attribute, should it be named just `@TransactionTimeout`  in that case?

Reply:
@Sanne while it's an int, we have a set a default value of `0` which means use the global tx time as per the TM API, so we are good. Anyways, you can't set null in an annotation :)

Reply:
Yeah, I agree with Sanne, I would throw an exception to begin with.

We can collect feedback from the field and see if we need to relax it later.

Reply:
Alright, @loicmathieu that looks like the last chance to throw an exception upon misuse.

Reply:
@emmanuelbernard no please don't give `0` a special meaning, it already has a special meaning in some related APIs. At least in JPA it translates to "lock without wait" on operations, so one might want to _intentionally_ set it to zero.

Reply:
@Sanne We are not giving it meaning, it is already given by https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionManager.html#setTransactionTimeout-int-

> seconds - The value of the timeout in seconds. If the value is zero, the transaction service restores the default value. If the value is negative a SystemException is thrown.

So we could give negative values our own meaning and adapt or MAX_INT. But 0 seems aligned with the spec.

Reply:
But that's what I'm referring to: since zero has a special meaning (in this case "restore the default value"), there is no way to use this annotation to set some different option and leave the transaction timeout "unaffected" by the annotation.

Which might be ok, I'm just warning that this limits the possibilities to reuse this annotation.

If the intent is to just have "timeout" on this `@TransactionConfig` then no problem, but perhaps it would be better to name it `@TransactionTimeout` - it was my understanding that we previously setted on the more generic name so to have more attributes in the future.

Reply:
@emmanuelbernard you want me to throw an exception in case the annotation is set in a nested method used by another method that already have the `@Transactional` annotation right ?

Reading the code, as the interceptor use `tm.begging()` and the Javadoc of this method said:
> Create a new transaction and associate it with the current thread.
> @exception NotSupportedException Thrown if the thread is already associated with a transaction and the Transaction Manager implementation does not support nested transactions.

The current implementation should throw an exception ... but I tested it and it didn't throw an exception ( something should have set allowSubTransaction on the Narayana TM because by default it should throw). So if we throw it in case of duplicated `@TransactionConfiguration` we will be stricted than what is done currently ... Anyway I'm OK with it but **I don't know who to check that we are in a nested method call.**

@Sanne regarding the value of `0` by default, for me the issue it that I reset everytime the TM to `0` in the current implementation, maybe using `-1` as a default value and doing nothing with it will be better. `0` is convenient because we can set it safely on the TM ...

Reply:
@loicmathieu about -1, create a UNSET_TIMEOUT constant. At least it will be understandable.

Reply:
@loicmathieu about the exception, you need to check but if I understand the code correctly, the exception would be raised in https://github.com/quarkusio/quarkus/blob/3ea3c82c07bd4c7c622f7402da1f9fef0236ce53/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TransactionalInterceptorBase.java#L247
In `invokeInCallerTx`

Reply:
As you say, something must have [setSupportSubtransactions](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/jta/common/JTAEnvironmentBean.java#L144) as the default is false, [this](https://github.com/quarkusio/quarkus/blob/30a6e37d30aca0122be5a7fca71b892d7ebe817c/extensions/narayana-jta/runtime/pom.xml#L43) pulls in narayana-jta and the [configuration file](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/narayana-jta/src/main/resources/jbossts-properties.xml) of narayana-jta does not override the default.

[Here](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/BaseTransaction.java#L75) is where a failed check throws a NotSupportedException as per the [standard](https://docs.oracle.com/javaee/7/api/javax/transaction/UserTransaction.html#begin--).

To check if transaction is already running you can just check [getStatus](https://docs.oracle.com/javaee/7/api/javax/transaction/UserTransaction.html#getStatus--).

Reply:
@tomjenkinson I don't this we are talking about subtransactions here. Just properly implementing https://docs.oracle.com/javaee/7/api/javax/transaction/Transactional.TxType.html
I think these are different beasts but correct me if I'm wrong.

Reply:
@emmanuelbernard @Sanne I use -1 instead or 0 and I create a constant UNSET_TRANSACTION in the @TransactionConfiguration annotation for holding it.

@tomjenkinson I throw a RuntimeException when the annotation is set on a method where the transaction has the status `TRANSACTION_NO_TRANSACTION`, the guide clearly state that the annotation needs to be present on the entry method of the transaction, I also add a line to the JavaDoc of the annotation that state this clearly.

Reply:
@emmanuelbernard I was referring to the report of @loicmathieu who said that Narayana should throw an exception if tm.begin is called and a TX was already on the thread and he tested it and it didn't throw an exception. @loicmathieu if you can tell me how you saw that I can try to run it here in the debugger?